File: chanlib/src/main/java/com/mimireader/chanlib/models/ChanThread.java
/*
 * Copyright (c) 2016. Eli Connelly
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.mimireader.chanlib.models;


import android.text.TextUtils;

import com.google.gson.annotations.Expose;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ChanThread {
    @Expose protected List<ChanPost> posts = new ArrayList<>();
    @Expose protected String boardName;
    @Expose protected String boardTitle;
    @Expose protected long threadId;

    public ChanThread() {
        this.threadId = -1;
    }

    public ChanThread(String boardName, long threadId, List<ChanPost> posts) {
        this.boardName = boardName;
        this.threadId = threadId;
        this.posts.addAll(posts);
    }

    public ChanThread(ChanThread thread) {
        this.boardName = thread.getBoardName();
        this.boardTitle = thread.getBoardTitle();
        this.threadId = thread.getThreadId();
        this.posts.addAll(thread.getPosts());
    }

    public List<ChanPost> getPosts() {
        return posts;
    }

    public void setPosts(List<ChanPost> posts) {
        this.posts = posts;
    }

    public String getBoardName() {
        return boardName;
    }

    public void setBoardName(String boardName) {
        this.boardName = boardName;
    }

    public long getThreadId() {
        return threadId;
    }

    public void setThreadId(long threadId) {
        this.threadId = threadId;
    }

    public String getBoardTitle() {
        return boardTitle;
    }

    public void setBoardTitle(String boardTitle) {
        this.boardTitle = boardTitle;
    }

    public static ChanThread empty() {
        return new ChanThread("", -1, Collections.<ChanPost>emptyList());
    }

    public static boolean isEmpty(ChanThread thread) {
        return thread == null || (TextUtils.isEmpty(thread.boardName) && thread.threadId == -1 && thread.posts.size() == 0);
    }
}

File: chanlib/src/main/java/com/mimireader/chanlib/models/ChanThread.kt
package com.mimireader.chanlib.models

import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import com.google.gson.annotations.Expose
import java.util.*

open class ChanThread : Parcelable {

    @Expose
    var posts: MutableList<ChanPost> = ArrayList()

    @Expose
    var boardName: String = ""

    @Expose
    var boardTitle: String = ""

    @Expose
    var threadId: Long = -1

    constructor(parcel: Parcel) : this() {
        boardName = parcel.readString() ?: ""
        boardTitle = parcel.readString() ?: ""
        threadId = parcel.readLong()
    }

    constructor() {
        threadId = -1
    }

    constructor(boardName: String, threadId: Long, posts: List<ChanPost>) : this() {
        this.boardName = boardName
        this.threadId = threadId
        this.posts.addAll(posts)
    }

    constructor(thread: ChanThread) : this() {
        boardName = thread.boardName
        boardTitle = thread.boardTitle
        threadId = thread.threadId
        posts.addAll(thread.posts)
    }

    override fun toString(): String {
        return "ChanThread{" +
                "board='" + boardName + '\'' +
                ", title='" + boardTitle + '\'' +
                ", thread id='" + threadId + '\'' +
                ", post count='" + posts.size + '\'' +
                '}'
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(boardName)
        parcel.writeString(boardTitle)
        parcel.writeLong(threadId)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object {

        @JvmField
        val CREATOR : Parcelable.Creator<ChanThread> = object : Parcelable.Creator<ChanThread> {
            override fun createFromParcel(parcel: Parcel): ChanThread {
                return ChanThread(parcel)
            }

            override fun newArray(size: Int): Array<ChanThread?> {
                return arrayOfNulls(size)
            }
        }

        @JvmStatic
        fun empty(): ChanThread {
            return ChanThread("", -1, emptyList<ChanPost>())
        }

        @JvmStatic
        fun isEmpty(thread: ChanThread): Boolean {
            return TextUtils.isEmpty(thread.boardName) && thread.threadId == -1L && thread.posts.size == 0
        }
    }
}